pkg/util/pool/pool.go

```diff
@@ -176,3 +176,71 @@ func (b *BatchBytes) Get(sz int) ([]byte, error) {
 	*slab = (*slab)[:len(*slab)+sz]
 	return (*slab)[len(*slab)-sz : len(*slab) : len(*slab)], nil
 }
+
+// SlabPool is NOT concurrency safe.
+type SlabPool[T any] struct {
+	delegate Interface
+	slabSize int
+	slabs    []*[]T
+}
+
+func NewSlabPool[T any](delegate Interface, slabSize int) *SlabPool[T] {
+	return &SlabPool[T]{
+		delegate: delegate,
+		slabSize: slabSize,
+	}
+}
+
+// Release all slices returned by Get. It's unsafe to access slices previously returned by Get
+// after calling Release().
+func (b *SlabPool[T]) Release() {
+	for _, slab := range b.slabs {
+		// The slab length will be reset to 0 in the Get().
+		b.delegate.Put(slab)
+	}
+
+	b.slabs = b.slabs[:0]
+}
+
+// Get returns a slice of T with the given length and capacity (both matches).
+func (b *SlabPool[T]) Get(size int) []T {
+	const lookback = 3
+
+	if size <= 0 {
+		return nil
+	}
+
+	// If the requested size is bigger than the slab size, then the slice
+	// can't be handled by this pool and will be allocated outside it.
+	if size > b.slabSize {
+		return make([]T, size)
```

Comment:
Nothing prevents us from doing it. My reasoning was that if we return it to the pool and then it will be reused by a request that needs a way smaller slice (e.g. few chunks) then there would be a lot of wasted memory utilization, while this memory could have been freed if it wasn't put back into the pool.

For the specific use case of this PR, having `size > b.slabSize` is very unlikely to happen.

Reply:
i saw your estimations, they make sense to me. They don't account for users that run with blocks larger than 24h. Or maybe blocks imported from thanos or prometheus, where they have blocks of up to 31d.

But then I ran some back of the envelope calculations - with prometheus' 31d blocks and 30s scrape interval, you'd get 744 chunks per block, which still fits in the slab size. I think throwing this away makes sense.

We can also see if we're hitting this case often in profiles